3.8.0: `filelocksqlite` is broken PyPy3.11

After upgrading to 3.8.0, I'm seeing lots of new test failures on PyPy3.11 7.3.20:

```pytb
$ pytest -k 'not postgres'  --maxfail=5
========================================================= test session starts =========================================================
platform linux -- Python 3.11.13[pypy-7.3.20-final], pytest-8.4.1, pluggy-1.6.0
rootdir: /tmp/PyrateLimiter
configfile: pyproject.toml
plugins: xdist-3.8.0, asyncio-1.1.0
asyncio: mode=Mode.STRICT, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 749 items / 119 deselected / 630 selected                                                                                   

tests/test_bucket_all.py ....F....F....F....F....F

============================================================== FAILURES ===============================================================
_________________________________________ test_bucket_01[clock0-create_filelocksqlite_bucket] _________________________________________

clock = <pyrate_limiter.clocks.MonotonicClock object at 0x00007f3781f03830>
create_bucket = <function create_filelocksqlite_bucket at 0x00007f3781fd56a0>

    @pytest.mark.asyncio
    async def test_bucket_01(clock: ClockSet, create_bucket):
        rates = [Rate(20, 1000)]
        bucket = BucketAsyncWrapper(await create_bucket(rates))
        assert bucket is not None
    
        peek = await bucket.peek(0)
        assert peek is None
    
>       await bucket.put(RateItem("my-item", await get_now(clock)))

tests/test_bucket_all.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pyrate_limiter/abstracts/wrappers.py:20: in put
    result = self.bucket.put(item)
             ^^^^^^^^^^^^^^^^^^^^^
pyrate_limiter/buckets/sqlite_bucket.py:119: in put
    self.conn.commit()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_sqlite3.Connection object at 0x00007f377e52b718>

    def commit(self):
        self._check_thread()
        self._check_closed()
        if not self.in_transaction:
            return
    
        # PyPy fix for non-refcounting semantics: since 2.7.13 (and in
        # <= 2.6.x), the statements are not automatically reset upon
        # commit.  However, if this is followed by some specific SQL
        # operations like "drop table", these open statements come in
        # the way and cause the "drop table" to fail.  On CPython the
        # problem is much less important because typically all the old
        # statements are freed already by reference counting.  So here,
        # we copy all the still-alive statements to another list which
        # is usually ignored, except if we get SQLITE_LOCKED
        # afterwards---at which point we reset all statements in this
        # list.
        self.__statements_already_committed = self.__statements[:]
    
        statement_star = _ffi.new('sqlite3_stmt **')
        ret = _lib.sqlite3_prepare_v2(self._db, b"COMMIT", -1,
                                      statement_star, _ffi.NULL)
        try:
            if ret != _lib.SQLITE_OK:
                raise self._get_exception(ret)
            ret = _lib.sqlite3_step(statement_star[0])
            if ret != _lib.SQLITE_DONE:
>               raise self._get_exception(ret)
E               _sqlite3.OperationalError: cannot commit transaction - SQL statements in progress

/usr/lib/pypy3.11/_sqlite3.py:673: OperationalError
-------------------------------------------------------- Captured stderr call ---------------------------------------------------------
INFO:pyrate_limiter:SQLite db path: /tmp/pyrate_limiter_0mEPt.sqlite
---------------------------------------------------------- Captured log call ----------------------------------------------------------
INFO     pyrate_limiter:conftest.py:87 SQLite db path: /tmp/pyrate_limiter_0mEPt.sqlite
_________________________________________ test_bucket_01[clock1-create_filelocksqlite_bucket] _________________________________________

clock = <pyrate_limiter.clocks.TimeClock object at 0x00007f3781f03868>
create_bucket = <function create_filelocksqlite_bucket at 0x00007f3781fd56a0>

    @pytest.mark.asyncio
    async def test_bucket_01(clock: ClockSet, create_bucket):
        rates = [Rate(20, 1000)]
        bucket = BucketAsyncWrapper(await create_bucket(rates))
        assert bucket is not None
    
        peek = await bucket.peek(0)
        assert peek is None
    
>       await bucket.put(RateItem("my-item", await get_now(clock)))

tests/test_bucket_all.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pyrate_limiter/abstracts/wrappers.py:20: in put
    result = self.bucket.put(item)
             ^^^^^^^^^^^^^^^^^^^^^
pyrate_limiter/buckets/sqlite_bucket.py:119: in put
    self.conn.commit()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_sqlite3.Connection object at 0x00007f377f11e170>

    def commit(self):
        self._check_thread()
        self._check_closed()
        if not self.in_transaction:
            return
    
        # PyPy fix for non-refcounting semantics: since 2.7.13 (and in
        # <= 2.6.x), the statements are not automatically reset upon
        # commit.  However, if this is followed by some specific SQL
        # operations like "drop table", these open statements come in
        # the way and cause the "drop table" to fail.  On CPython the
        # problem is much less important because typically all the old
        # statements are freed already by reference counting.  So here,
        # we copy all the still-alive statements to another list which
        # is usually ignored, except if we get SQLITE_LOCKED
        # afterwards---at which point we reset all statements in this
        # list.
        self.__statements_already_committed = self.__statements[:]
    
        statement_star = _ffi.new('sqlite3_stmt **')
        ret = _lib.sqlite3_prepare_v2(self._db, b"COMMIT", -1,
                                      statement_star, _ffi.NULL)
        try:
            if ret != _lib.SQLITE_OK:
                raise self._get_exception(ret)
            ret = _lib.sqlite3_step(statement_star[0])
            if ret != _lib.SQLITE_DONE:
>               raise self._get_exception(ret)
E               _sqlite3.OperationalError: cannot commit transaction - SQL statements in progress

/usr/lib/pypy3.11/_sqlite3.py:673: OperationalError
-------------------------------------------------------- Captured stderr call ---------------------------------------------------------
INFO:pyrate_limiter:SQLite db path: /tmp/pyrate_limiter_p7Zu2.sqlite
---------------------------------------------------------- Captured log call ----------------------------------------------------------
INFO     pyrate_limiter:conftest.py:87 SQLite db path: /tmp/pyrate_limiter_p7Zu2.sqlite
_________________________________________ test_bucket_01[clock2-create_filelocksqlite_bucket] _________________________________________

clock = <pyrate_limiter.clocks.TimeAsyncClock object at 0x00007f3781f038a0>
create_bucket = <function create_filelocksqlite_bucket at 0x00007f3781fd56a0>

    @pytest.mark.asyncio
    async def test_bucket_01(clock: ClockSet, create_bucket):
        rates = [Rate(20, 1000)]
        bucket = BucketAsyncWrapper(await create_bucket(rates))
        assert bucket is not None
    
        peek = await bucket.peek(0)
        assert peek is None
    
>       await bucket.put(RateItem("my-item", await get_now(clock)))

tests/test_bucket_all.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pyrate_limiter/abstracts/wrappers.py:20: in put
    result = self.bucket.put(item)
             ^^^^^^^^^^^^^^^^^^^^^
pyrate_limiter/buckets/sqlite_bucket.py:119: in put
    self.conn.commit()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_sqlite3.Connection object at 0x000055a905a84368>

    def commit(self):
        self._check_thread()
        self._check_closed()
        if not self.in_transaction:
            return
    
        # PyPy fix for non-refcounting semantics: since 2.7.13 (and in
        # <= 2.6.x), the statements are not automatically reset upon
        # commit.  However, if this is followed by some specific SQL
        # operations like "drop table", these open statements come in
        # the way and cause the "drop table" to fail.  On CPython the
        # problem is much less important because typically all the old
        # statements are freed already by reference counting.  So here,
        # we copy all the still-alive statements to another list which
        # is usually ignored, except if we get SQLITE_LOCKED
        # afterwards---at which point we reset all statements in this
        # list.
        self.__statements_already_committed = self.__statements[:]
    
        statement_star = _ffi.new('sqlite3_stmt **')
        ret = _lib.sqlite3_prepare_v2(self._db, b"COMMIT", -1,
                                      statement_star, _ffi.NULL)
        try:
            if ret != _lib.SQLITE_OK:
                raise self._get_exception(ret)
            ret = _lib.sqlite3_step(statement_star[0])
            if ret != _lib.SQLITE_DONE:
>               raise self._get_exception(ret)
E               _sqlite3.OperationalError: cannot commit transaction - SQL statements in progress

/usr/lib/pypy3.11/_sqlite3.py:673: OperationalError
-------------------------------------------------------- Captured stderr call ---------------------------------------------------------
INFO:pyrate_limiter:SQLite db path: /tmp/pyrate_limiter_3rMjM.sqlite
---------------------------------------------------------- Captured log call ----------------------------------------------------------
INFO     pyrate_limiter:conftest.py:87 SQLite db path: /tmp/pyrate_limiter_3rMjM.sqlite
_________________________________________ test_bucket_02[clock0-create_filelocksqlite_bucket] _________________________________________

clock = <pyrate_limiter.clocks.MonotonicClock object at 0x00007f3781f03830>
create_bucket = <function create_filelocksqlite_bucket at 0x00007f3781fd56a0>

    @pytest.mark.asyncio
    async def test_bucket_02(clock: ClockSet, create_bucket):
        rates = [Rate(30, 1000), Rate(50, 2000)]
        bucket = BucketAsyncWrapper(await create_bucket(rates))
        start = time()
    
        while await bucket.count() < 150:
>           await bucket.put(RateItem("item", await get_now(clock)))

tests/test_bucket_all.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pyrate_limiter/abstracts/wrappers.py:20: in put
    result = self.bucket.put(item)
             ^^^^^^^^^^^^^^^^^^^^^
pyrate_limiter/buckets/sqlite_bucket.py:119: in put
    self.conn.commit()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_sqlite3.Connection object at 0x00007f377f20ab10>

    def commit(self):
        self._check_thread()
        self._check_closed()
        if not self.in_transaction:
            return
    
        # PyPy fix for non-refcounting semantics: since 2.7.13 (and in
        # <= 2.6.x), the statements are not automatically reset upon
        # commit.  However, if this is followed by some specific SQL
        # operations like "drop table", these open statements come in
        # the way and cause the "drop table" to fail.  On CPython the
        # problem is much less important because typically all the old
        # statements are freed already by reference counting.  So here,
        # we copy all the still-alive statements to another list which
        # is usually ignored, except if we get SQLITE_LOCKED
        # afterwards---at which point we reset all statements in this
        # list.
        self.__statements_already_committed = self.__statements[:]
    
        statement_star = _ffi.new('sqlite3_stmt **')
        ret = _lib.sqlite3_prepare_v2(self._db, b"COMMIT", -1,
                                      statement_star, _ffi.NULL)
        try:
            if ret != _lib.SQLITE_OK:
                raise self._get_exception(ret)
            ret = _lib.sqlite3_step(statement_star[0])
            if ret != _lib.SQLITE_DONE:
>               raise self._get_exception(ret)
E               _sqlite3.OperationalError: cannot commit transaction - SQL statements in progress

/usr/lib/pypy3.11/_sqlite3.py:673: OperationalError
-------------------------------------------------------- Captured stderr call ---------------------------------------------------------
INFO:pyrate_limiter:SQLite db path: /tmp/pyrate_limiter_RcJrH.sqlite
---------------------------------------------------------- Captured log call ----------------------------------------------------------
INFO     pyrate_limiter:conftest.py:87 SQLite db path: /tmp/pyrate_limiter_RcJrH.sqlite
_________________________________________ test_bucket_02[clock1-create_filelocksqlite_bucket] _________________________________________

clock = <pyrate_limiter.clocks.TimeClock object at 0x00007f3781f03868>
create_bucket = <function create_filelocksqlite_bucket at 0x00007f3781fd56a0>

    @pytest.mark.asyncio
    async def test_bucket_02(clock: ClockSet, create_bucket):
        rates = [Rate(30, 1000), Rate(50, 2000)]
        bucket = BucketAsyncWrapper(await create_bucket(rates))
        start = time()
    
        while await bucket.count() < 150:
>           await bucket.put(RateItem("item", await get_now(clock)))

tests/test_bucket_all.py:67: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pyrate_limiter/abstracts/wrappers.py:20: in put
    result = self.bucket.put(item)
             ^^^^^^^^^^^^^^^^^^^^^
pyrate_limiter/buckets/sqlite_bucket.py:119: in put
    self.conn.commit()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_sqlite3.Connection object at 0x00007f37816a4e20>

    def commit(self):
        self._check_thread()
        self._check_closed()
        if not self.in_transaction:
            return
    
        # PyPy fix for non-refcounting semantics: since 2.7.13 (and in
        # <= 2.6.x), the statements are not automatically reset upon
        # commit.  However, if this is followed by some specific SQL
        # operations like "drop table", these open statements come in
        # the way and cause the "drop table" to fail.  On CPython the
        # problem is much less important because typically all the old
        # statements are freed already by reference counting.  So here,
        # we copy all the still-alive statements to another list which
        # is usually ignored, except if we get SQLITE_LOCKED
        # afterwards---at which point we reset all statements in this
        # list.
        self.__statements_already_committed = self.__statements[:]
    
        statement_star = _ffi.new('sqlite3_stmt **')
        ret = _lib.sqlite3_prepare_v2(self._db, b"COMMIT", -1,
                                      statement_star, _ffi.NULL)
        try:
            if ret != _lib.SQLITE_OK:
                raise self._get_exception(ret)
            ret = _lib.sqlite3_step(statement_star[0])
            if ret != _lib.SQLITE_DONE:
>               raise self._get_exception(ret)
E               _sqlite3.OperationalError: cannot commit transaction - SQL statements in progress

/usr/lib/pypy3.11/_sqlite3.py:673: OperationalError
-------------------------------------------------------- Captured stderr call ---------------------------------------------------------
INFO:pyrate_limiter:SQLite db path: /tmp/pyrate_limiter_COq1M.sqlite
---------------------------------------------------------- Captured log call ----------------------------------------------------------
INFO     pyrate_limiter:conftest.py:87 SQLite db path: /tmp/pyrate_limiter_COq1M.sqlite
======================================================= short test summary info =======================================================
FAILED tests/test_bucket_all.py::test_bucket_01[clock0-create_filelocksqlite_bucket] - _sqlite3.OperationalError: cannot commit transaction - SQL statements in progress
FAILED tests/test_bucket_all.py::test_bucket_01[clock1-create_filelocksqlite_bucket] - _sqlite3.OperationalError: cannot commit transaction - SQL statements in progress
FAILED tests/test_bucket_all.py::test_bucket_01[clock2-create_filelocksqlite_bucket] - _sqlite3.OperationalError: cannot commit transaction - SQL statements in progress
FAILED tests/test_bucket_all.py::test_bucket_02[clock0-create_filelocksqlite_bucket] - _sqlite3.OperationalError: cannot commit transaction - SQL statements in progress
FAILED tests/test_bucket_all.py::test_bucket_02[clock1-create_filelocksqlite_bucket] - _sqlite3.OperationalError: cannot commit transaction - SQL statements in progress
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 5 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================= 5 failed, 20 passed, 119 deselected in 89.59s (0:01:29) =======================================
```

I've cut the test run at 5 failures, but all tests fail with the same error.

```console
$ uv pip list
Package           Version Editable project location
----------------- ------- -------------------------
execnet           2.1.1
filelock          3.18.0
iniconfig         2.1.0
packaging         25.0
pluggy            1.6.0
psycopg-pool      3.2.6
pygments          2.19.2
pyrate-limiter    3.8.0   /tmp/PyrateLimiter
pytest            8.4.1
pytest-asyncio    1.1.0
pytest-xdist      3.8.0
redis             6.2.0
typing-extensions 4.14.1
```

--- a/pyrate_limiter/buckets/sqlite_bucket.py
+++ b/pyrate_limiter/buckets/sqlite_bucket.py
@@ -99,7 +99,9 @@ class SQLiteBucket(AbstractBucket):
     def put(self, item: RateItem) -> bool:
         with self.lock:
             query, parameters = self._build_full_count_query(item.timestamp)
-            rate_limit_counts = self.conn.execute(query, parameters).fetchall()
+            cur = self.conn.execute(query, parameters)
+            rate_limit_counts = cur.fetchall()
+            cur.close()
 
             for idx, result in enumerate(rate_limit_counts):
                 interval, count = result
@@ -115,7 +117,7 @@ class SQLiteBucket(AbstractBucket):
                 [f"('{name}', {item.timestamp})" for name in [item.name] * item.weight]
             )
             query = (Queries.PUT_ITEM.format(table=self.table)) % items
-            self.conn.execute(query)
+            self.conn.execute(query).close()
             self.conn.commit()
             return True
 
@@ -128,28 +130,35 @@ class SQLiteBucket(AbstractBucket):
                 interval=self.rates[-1].interval,
                 current_timestamp=current_timestamp,
             )
-            count = self.conn.execute(query).fetchone()[0]
+            cur = self.conn.execute(query)
+            count = cur.fetchone()[0]
             query = Queries.LEAK.format(table=self.table, count=count)
-            self.conn.execute(query)
+            cur.execute(query)
+            cur.close()
             self.conn.commit()
             return count
 
     def flush(self) -> None:
         with self.lock:
-            self.conn.execute(Queries.FLUSH.format(table=self.table))
+            self.conn.execute(Queries.FLUSH.format(table=self.table)).close()
             self.conn.commit()
             self.failing_rate = None
 
     def count(self) -> int:
         with self.lock:
-            return self.conn.execute(
+            cur = self.conn.execute(
                 Queries.COUNT_ALL.format(table=self.table)
-            ).fetchone()[0]
+            )
+            ret = cur.fetchone()[0]
+            cur.close()
+            return ret
 
     def peek(self, index: int) -> Optional[RateItem]:
         with self.lock:
             query = Queries.PEEK.format(table=self.table, count=index)
-            item = self.conn.execute(query).fetchone()
+            cur = self.conn.execute(query)
+            item = cur.fetchone()
+            cur.close()
 
             if not item:
                 return None
@@ -195,11 +204,12 @@ class SQLiteBucket(AbstractBucket):
                 check_same_thread=False,
             )
 
+            cur = sqlite_connection.cursor()
             if use_file_lock:
-                sqlite_connection.execute("PRAGMA journal_mode=WAL;")
+                cur.execute("PRAGMA journal_mode=WAL;")
 
             if create_new_table:
-                sqlite_connection.execute(
+                cur.execute(
                     Queries.CREATE_BUCKET_TABLE.format(table=table)
                 )
 
@@ -208,7 +218,8 @@ class SQLiteBucket(AbstractBucket):
                 table_name=table,
             )
 
-            sqlite_connection.execute(create_idx_query)
+            cur.execute(create_idx_query)
+            cur.close()
             sqlite_connection.commit()
 
             return cls(rates, sqlite_connection, table=table, lock=file_lock)

--- a/pyrate_limiter/clocks.py
+++ b/pyrate_limiter/clocks.py
@@ -49,7 +49,9 @@ class SQLiteClock(AbstractClock):
         return cls(conn)
 
     def now(self) -> int:
-        now = self.conn.execute(self.time_query).fetchone()[0]
+        cur = self.conn.execute(self.time_query)
+        now = cur.fetchone()[0]
+        cur.close()
         return int(now)
 
 

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -96,8 +96,9 @@ async def create_sqlite_bucket(rates: List[Rate], file_lock: bool = False):
     drop_table_query = Queries.DROP_TABLE.format(table=table_name)
     drop_index_query = Queries.DROP_INDEX.format(index=index_name)
 
-    conn.execute(drop_table_query)
-    conn.execute(drop_index_query)
+    cur = conn.execute(drop_table_query)
+    cur.execute(drop_index_query)
+    cur.close()
     conn.commit()
 
     bucket = SQLiteBucket.init_from_file(

